Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -203,146 +203,87 @@ function xASL_wrp_ResampleASL(x)
-        
-        % Save single PWI
-        fprintf('%s\n', PathPWI{iSpace});
-        xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-        
-    else
-        % Paired subtraction
-        [ControlIm, LabelIm] = xASL_quant_GetControlLabelOrder(ASL_im);
-        ASL_im = ControlIm - LabelIm;
-
-        % Average PWI - multiPLD
-        if x.modules.asl.bMultiPLD 
-			% Skip every other value in x.Q.Initial_PLD as it was stored for both control and label images 
-			% and we need the PLD vector now for the pairwise subtractions only
-			Initial_PLD_PWI = x.Q.Initial_PLD(1:2:end);
-
-			% The labeling durations also need to be taken into account and unique combinations of PLDs with LDs should be considered
-			if numel(x.Q.Initial_PLD) == numel(x.Q.LabelingDuration)
-				Initial_LabDur_PWI = x.Q.LabelingDuration(1:2:end);
-			else
-				Initial_LabDur_PWI = ones(size(Initial_PLD_PWI)) * x.Q.LabelingDuration;
-			end
-        
-			% After averaging across PLDs, we'll obtain these unique PLDs+LD combinations
-			% indexNew lists for each original position to where it should be averaged
-			[~, ~, indexNew] = unique([Initial_PLD_PWI, Initial_LabDur_PWI], 'stable', 'rows');
-			
-            % MultiPLD PWI after averaging
-			PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), max(indexNew)); % preallocate PWI
-            for nPLD = 1:max(indexNew)
-                PWI(:, :, :, nPLD) = xASL_stat_MeanNan(ASL_im(:, :, :, indexNew == nPLD), 4); % Averaged PWI4D 
-            end
-            
-            % Save PWI
-            fprintf('%s\n', PathPWI4D{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI4D{iSpace}, PWI, 32, false);
-            
-            PWIsingle = xASL_stat_MeanNan(ASL_im, 4); % create single PWI for further steps in ASL module
-            
-            % Save single PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-            
-        else
-            PWI = xASL_stat_MeanNan(ASL_im, 4); % singlePLD PWI
-            % Save PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWI, 32, false);
-        end            
+    % Here we reload this again, because we have different 
+    ASL4D = xASL_io_Nifti2Im(PathASL4D{iSpace}); % Load time-series nifti
+    
+    if numel(size(ASL4D))>4
+        warning('In BIDS ASL NIfTIs should have [X Y Z n/PLD/TE/etc] as 4 dimensions');
+        error(['ASL4D NIfTI has more than 4 dimensions: ' PathASL4D{iSpace}]);
     end
+
+    [PWI, PWI3D, PWI4D, x] = xASL_im_ASLSubtractionAveraging(x, ASL4D);
+    
+    
+    % Save subtracted/averaged to disk
+
+    % Save PWI4D (subtracted only)
+    fprintf('%s\n', PathPWI4D{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI4D{iSpace}, PWI4D, 32, false);
+    spm_jsonwrite(PathPWI4Djson{iSpace}, x.Q);

```

Comment:
Well if we want JSON sidecars for knowing the parameters in the PWI4D and PWI3D, we should be consistent (which is also according to BIDS). I'm also happy to do this for native space only though, but it will make the code very dirty. Currently, any PWI4D or PWI3D can be read, irregardless of the space it is in, and the extension replaced by json. That won't be the case if we save the sidecar space-dependent.

Reply:
I agree - all PWI3D and PWI4D should have their JSON files. But I strongly discourage saving PWI4D and ASL4D in the standard space at all! See my other comment about this.

Reply:
Yes, and my reasoning there. Let's keep this discussing this, but we can then close this here right. Because if we decide to not do this in standard space, we keep this here, but simply iterate over a single (native) space only. So can you resolve this conversation?

Reply:
We can close this here. But what about the PWI4D in standard space that is taking a lot of space?